### PR TITLE
Fix graded buoyancy problems

### LINKED ISSUE #1296: Buoyancy restoring calculations are wrong when using graded buoyancy.
## Environment
* OS Version: All
* Fortress and Garden.

## Description
Given an example world with restoring moments uniform buoyancy and graded buoyancy provide very different results.

## Steps to reproduce

I have attached a simple world here:
https://gist.github.com/arjo129/ce5e553407163f32572bbfd2dc990740

This world has a setup where we have the center of buoyancy below the center of gravity.
buoyancy_volumes
As one can see the center of buoyancy (yellow being the buoyancy mesh) is below the actual center of the box which in turn is the center of gravity. 

When using graded buoyancy the world seems stable.
buoyancy_graded_restoring

In practice according to [common knowledge](https://www.marineinsight.com/naval-architecture/understanding-stability-submarine/) the center of buoyancy of a submarine should always be above the center of gravity for the world should be unstable:
image

This is consistent with the behavior of uniform buoyancy:
buoyancy_normal_restoring

### PR DESCRIPTION
**Pull request by arjo129**
# 🦟 Bug fix

Fixes #1296 and #1210

## Summary
The primary target of this PR is #1296. This PR changes an incorrect sign line 224 of the buoyancy plugin when it was running in graded buoyancy mode. Essentially when running the example in #1296 now both graded and uniform mode are unstable. This is the correct behavior. For a stable example see the added tests in this PR. In particular both SDFs should be stable and have small oscillations, but previously the graded buoyancy example would have kept on spinning.

Additionally #1210 would mean the graded buoyancy test would keep sinking.  This PR removes the need for checking the `newPose` as we nolonger need the inertial pose. The reason I was initially using the inertialpose was because I had thought that `Link::AddWorldWrench` applies a force at the center of mass, but in reality the original behaviour in uniform buoyancy was correct as `Link::AddWorldWrench` applies a force at the link itself.

### TL;DR
Before times when running https://github.com/ignitionrobotics/ign-gazebo/blob/1d461b849b92104299bc03879bb67db368fe10c9/test/worlds/buoyancy_graded_restoring_moments.sdf : 

buoyancy_before_times

After times:
buoyancy_after_times

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [ ] ~Updated documentation (as needed)~
- [ ] ~Updated migration guide (as needed)~
- [x] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [x] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

**Comment by chapulina** [comment]
I think there may still be some offset calculations that are incorrect, see

* https://github.com/ignitionrobotics/ign-gazebo/pull/1301

File: test/integration/buoyancy.cc [hunk 1]
```diff
@@ -42,6 +42,121 @@ class BuoyancyTest : public InternalFixture<::testing::Test>
 {
 };
 
+/////////////////////////////////////////////////
+TEST_F(BuoyancyTest, IGN_UTILS_TEST_DISABLED_ON_WIN32(RestoringMoments))
+{
+  // This test checks if the restoring moments are correctly calculated accross
+  // both uniform and graded modes when the vehicle is fully submerged.
+  std::size_t iterations{10000};
+  std::vector<math::Pose3d> poses_uniform, poses_graded;
```
Comment (chapulina) [inline test/integration/buoyancy.cc:51]:
```suggestion
  std::vector<math::Pose3d> posesUniform, posesGraded;
```

Style nit. There are more instances throughout the test.

File: test/integration/buoyancy.cc [hunk 2]
```diff
@@ -42,6 +42,121 @@ class BuoyancyTest : public InternalFixture<::testing::Test>
...
+  {
+    // Start server
+    ServerConfig serverConfig;
+    const auto sdfFile = common::joinPaths(std::string(PROJECT_SOURCE_PATH),
+      "/test/worlds/buoyancy_uniform_restoring_moments.sdf");
```
Comment (chapulina) [inline test/integration/buoyancy.cc:56]: The segments need to be split

```suggestion
      "test", "worlds", "buoyancy_uniform_restoring_moments.sdf");
```

File: test/integration/buoyancy.cc [hunk 3]
```diff
@@ -42,6 +42,121 @@ class BuoyancyTest : public InternalFixture<::testing::Test>
...
+  }
+
+  double max_uniform_roll{0}, max_graded_roll{0};
+  double min_uniform_roll{0}, min_graded_roll{0};
+
+  ASSERT_EQ(poses_graded.size(), poses_uniform.size());
+  ASSERT_EQ(poses_uniform.size(), iterations);
+  for (std::size_t i = 0; i <  poses_graded.size(); i++)
+  {
+    // The two modes should track a similar course when fully submerged
```
Comment (chapulina) [inline test/integration/buoyancy.cc:119]:
```suggestion
    // The two models should track a similar course when fully submerged
```

File: test/integration/buoyancy.cc [hunk 4]
```diff
@@ -42,6 +42,121 @@ class BuoyancyTest : public InternalFixture<::testing::Test>
...
+    // Ignore roll for now as that changes a lot and the two sims are not
+    // perfectly in sync.
```
Comment (chapulina) [inline test/integration/buoyancy.cc:121]: Both worlds are using the same step size and run for the same number of iterations. What's missing for them to be in sync?

File: test/integration/buoyancy.cc [hunk 5]
```diff
@@ -42,6 +42,121 @@ class BuoyancyTest : public InternalFixture<::testing::Test>
 {
 };
 
+/////////////////////////////////////////////////
+TEST_F(BuoyancyTest, IGN_UTILS_TEST_DISABLED_ON_WIN32(RestoringMoments))
```
Comment (chapulina) [inline test/integration/buoyancy.cc:46]: Sweet, I think this is the first test where we actually start 2 separate servers and compare results, great job!

File: test/worlds/buoyancy_graded_restoring_moments.sdf [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+      name="ignition::gazebo::systems::Physics">
+    </plugin>
+    <plugin
+      filename="ignition-gazebo-user-commands-system"
+      name="ignition::gazebo::systems::UserCommands">
+    </plugin>
+    <plugin
+      filename="ignition-gazebo-scene-broadcaster-system"
+      name="ignition::gazebo::systems::SceneBroadcaster">
+    </plugin>
```
Comment (chapulina) [inline test/worlds/buoyancy_graded_restoring_moments.sdf:19]: We could strip `UserCommands`, `SceneBroadcaster` and the sun for the world for performance

File: test/worlds/buoyancy_graded_restoring_moments.sdf [hunk 7]
```diff
@@ -0,0 +1,79 @@
+<sdf version="1.6">
+  <world name="buoyancy">
+
+    <physics name="1ms" type="ode">
```
Comment (chapulina) [inline test/worlds/buoyancy_graded_restoring_moments.sdf:4]:
```suggestion
    <physics name="1ms" type="ignored">
```

Nitpick. We've been removing `ode` from all these worlds because it can be misleading

**Review by chapulina — COMMENTED** [review]
Besides https://github.com/ignitionrobotics/ign-gazebo/pull/1301, I just have some more minor comments here.

File: test/integration/buoyancy.cc (hunk 4 above)
Comment (arjo129) [reply test/integration/buoyancy.cc:121]: I'm not very sure but they do seem to diverge over time. I guess minor floating point errors build up? Again, this is something I'm speculating and not sure abot...

**Comment by chapulina** [comment]
I pushed d013573 to disable the buoyancy engine test that was introduced on #1298 on Windows. It's not directly related to this PR, but it's also "graded buoyancy", so I'm taking advantage of the CI run here.

**Review by chapulina — APPROVED** [review]
LGTM with clean CI. I think we should eventually look deeper into why the 2 servers on the test are giving slightly different results though.

**Comment by osrf-triage** [comment]
This pull request has been mentioned on **Gazebo Community**. There might be relevant details there:

https://community.gazebosim.org/t/new-ignition-releases-2022-03-01-citadel-edifice-fortress/1313/1
